<a href="https://colab.research.google.com/github/venkatanadikatla/pytorch/blob/main/Week_9_Summative.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Question 1

Given an 5 × 5 × 2 input
𝐼
I,calculate the result of the convolution with 3 × 3 × 2 filter
𝐾
K and bias = 1, with stride = 2 and no padding.

### Answer 1
Given:


*   Input I of size
5
×
5
×
2
*   Filter
𝐾
 of size
3
×
3
×
2

*   Bias = 1
*   Stride (s) = 2


*   No padding



### Input Matrices:

### Separating these two Input matrices into I1 and I2


\[I_1 =
\begin{bmatrix}
2 & 2 & 0 & 1 & 1 \\
2 & 1 & 0 & 1 & 0 \\
1 & 2 & 0 & 2 & 1 \\
2 & 0 & 1 & 0 & 0 \\
1 & 0 & 0 & 1 & 0 \\
\end{bmatrix}\]



\[I_2 = \begin{bmatrix}
1 & 1 & 0 & 2 & 2 \\
1 & 0 & 2 & 1 & 0 \\
1 & 2 & 2 & 1 & 2 \\
0 & 1 & 1 & 0 & 1 \\
0 & 0 & 1 & 2 & 2 \\
\end{bmatrix} \]

### Filter Matrices:

### Separating these two Filters into K1 and K2

\[K_1 = \begin{bmatrix}
-1 & -1 & 0 \\
-1 & -1 & 1 \\
0 & 1 & 1 \\
\end{bmatrix} \]

\[K_2 = \begin{bmatrix}
1 & 1 & 1 \\
1 & 0 & -1 \\
-1 & -1 & -1 \\
\end{bmatrix} \]


## Steps for Convolution:
### Determine the output dimensions:
The formula for the output dimension with no padding and stride

s is given by:

Output Dimension
= [(W-F)/s]+1

where W is the input dimension and F is the filter dimension.

For our case:

Output Dimension
=
[(5 - 3) / 2]+1 = [2 / 2] + 1 = 2 (Output Height and Width)

So the output will be
2 X 2 X 1

<br/>


### Steps for Convolution:

#### Top-left corner(0,0):

\[ Region from I_1 =
\begin{bmatrix}
2 & 2 & 0 \\
2 & 1 & 0 \\
1 & 2 & 0 \\
\end{bmatrix}\]



\[I_2 = \begin{bmatrix}
1 & 1 & 0\\
1 & 0 & 2\\
1 & 2 & 2 \\
\end{bmatrix} \]

Element-wise product sum:

For I1 and K1:

(2x-1)+(2x-1)+(0x0)+(2x-1)+(1x-1)+(0x1)+(1x0)+(2x1)+(0x1)
=-2-2+0-2-1+0+0+2+0 = -5

For I2 and K2:

(1x1)+(1x1)+(0x1)+(1x1)+(0x0)+(2x-1)+(1x-1)+(2x-1)+(2x-1)
=1+1+0+1+0-2-1-2-2 = -4

**Sum and add bias: -5+(-4)+1 = -8**

<br/>
<br/>

#### Top-right corner(0,2):

\[ Region from I_1 =
\begin{bmatrix}
0 & 1 & 1 \\
0 & 1 & 0 \\
0 & 2 & 1 \\
\end{bmatrix}\]


\[ Region from I_2 = \begin{bmatrix}
0 & 2 & 2 \\
2 & 1 & 0 \\
2 & 1 & 2 \\
\end{bmatrix} \]

Element-wise product sum:

For I1 and K1:

(0x-1)+(1x -1)+(1x0)+(0x-1)+(1x-1)+(0x1)+(0x0)+(2x1)+(1x1)

= 0 - 1 + 0 - 1 - 1 + 0 + 0 + 2 + 1 = 0

<br/>

For I2 and K2:

(0x1)+(2x1)+(2x1)+(2x1)+(1x0)+(0x-1)+(2x-1)+(1x-1)+(2x-1)

= 0 + 2 + 2 + 2 + 0 + 0 - 2 - 1 - 2 = 1

**Sum and add bias: 0+1+1 = 1**

<br/>
<br/>

#### Bottom-left corner(0,2):


\[Region from I_1 =
\begin{bmatrix}
1 & 2 & 0\\
2 & 0 & 1\\
1 & 0 & 0\\
\end{bmatrix}\]



\[Region from I_2 = \begin{bmatrix}
1 & 2 & 2\\
0 & 1 & 1\\
0 & 0 & 1\\
\end{bmatrix} \]

Element-wise product sum:

For I1 and K1:

(1x-1)+(2x-1)+(0x0)+(2x-1)+(0x-1)+(1x1)+(1x0)+(0x1)+(0x1)

= -1-2+0-2+0+1+0+0+0 = -4
<br/>
For I2 and K2:

(1x1)+(2x1)+(2x1)+(0x1)+(1x0)+(1x-1)+(0x-1)+(0x-1)+(1x-1)

= 1+2+2+0+0-1-0-0-1 = 3
<br/>
**Sum and add bias:-4+3+1 = 0**

#### Bottom-right corner(2,2):

\[Region from I_1 =
\begin{bmatrix}
0 & 2 & 1 \\
1 & 0 & 0 \\
0 & 1 & 0 \\
\end{bmatrix}\]


\[Region from I_2 = \begin{bmatrix}
2 & 1 & 2 \\
1 & 0 & 1 \\
1 & 2 & 2 \\
\end{bmatrix} \]


Element-wise product sum:

For I1 and K1:

(0x-1)+(2x-1)+(1x0)+(1x-1)+(0x-1)+(0x1)+(0x0)+(1x1)+(0x1)

= 0-2+0-1-0+0+0+1+0 = -2

For I2 and K2:

(2X1)+(1X1)+(2X1)+(1X1)+(0X0)+(1X-1)+(1X-1)+(2X-1)+(2X-1)

= 2+1+2+1+0-1-1-2-2 = 0

**Sum and add bias:-2+0+1 = -1**

<br/>
<br/>
<br/>

The output feature map is:
\[
\begin{bmatrix}
-8 & 1 \\
0 & -1 \\
\end{bmatrix}
\]


<b/>
<b/><b/>
<b/><b/>
<b/>
<b/>
<b/>
<b/><b/>
<b/><b/>
<b/>
<b/>
<b/>
<b/><b/>
<b/><b/>
<b/>
<b/>

### Question 2
Calculate the result of applying ReLU to the answer of (a).

In [10]:
# Calculating the ReLU output through using the code.
import torch
conv_output = torch.tensor([[-8,1], [0,1]])
relu_output = torch.relu(conv_output)
print(relu_output)

tensor([[0, 1],
        [0, 1]])


### Answering the question 2 in a LaTex format:

To apply the ReLU activation function to the result of the convolution, we simply replace any negative values with 0 as it the ReLU function itself calls for ReLU(z) =  max(0,z).

#### Given Convolution Ouput:
\[
\begin{bmatrix}
-8 & 1 \\
0 & -1 \\
\end{bmatrix}
\]

Applying ReLU:

\[
\begin{bmatrix}
ReLU(-8) & ReLU(1) \\
ReLU(0) & ReLU(-1) \\
\end{bmatrix}
\]

so, the result of applying ReLU to the conv ouput is:
\[
\begin{bmatrix}
0 & 1 \\
0 & 0 \\
\end{bmatrix}
\]




<br/>
<br/>
<br/>

# Question 3: Given an input I, calculate the result of the max pooling with a 2 × 2 filter, stride = 2.

\[
I = \begin{bmatrix}
17 & 14 & 21 & 0 & 30 & 32 \\
1 & 38 & 16 & 33 & 34 & 23 \\
12 & 7 & 26 & 18 & 5 & 28 \\
22 & 25 & 11 & 40 & 15 & 19 \\
13 & 36 & 24 & 48 & 3 & 20 \\
47 & 29 & 4 & 6 & 42 & 31 \\
\end{bmatrix}
\]

<br/>
<br/>

### In order to calculate the max pooling output, calculate the following output dimensions:


*   Given filter size 2X2 (This indicates that we will look 2X2 regions of the input matrix (I) to perform the max pooling)
*   Stride (s) = 2 (A stride of 2 means the filter moves 2 steps at a time)

### Output Dimension
= [(W-F)/s]+1

where W is the input dimension and F is the filter dimension.

For our case:

Output Dimension
=
[(6 - 2) / 2]+1 = [4 / 2] + 1 = 3 (Output Height and Width)

So the output will be
3 X 3 X 1

<br/>

### Max Pooling Calculation:

**Region (0,0) to (1,1)**
This region includes the elements from the first and second rows and the first and second columns:

\[\begin{bmatrix}
17 & 14 \\
1 & 38 \\
\end{bmatrix} \]
<br/>
**Max Value: 38**
<br/>
<br/>
**Region (0,2) to (1,3)**
This region includes the elements from the first and second rows and the third and fourth columns:
\[\begin{bmatrix}
21 & 0 \\
16 & 33 \\
\end{bmatrix} \]
<br/>
**Max Value: 33**
<br/>
<br/>
**Region (0,4) to (1,5)**
This region includes the elements from the first and second rows and the fifth and sixth columns:
\[\begin{bmatrix}
30 & 32 \\
34 & 23 \\
\end{bmatrix} \]
<br/>
**Max Value: 34**
<br/>
<br/>
**Region (2,0) to (3,1)**
This region includes the elements from the third and fourth rows and the first and second columns:
\[\begin{bmatrix}
12 & 7 \\
22 & 25 \\
\end{bmatrix} \]
<br/>
**Max Value: 25**
<br/>
<br/>
**Region (2,2) to (3,3)**
This region includes the elements from the third and fourth rows and the third and fourth columns:
\[\begin{bmatrix}
26 & 18 \\
11 & 40 \\
\end{bmatrix} \]
<br/>
**Max Value: 40**
<br/>
<br/>
**Region (2,4) to (3,5)**
This region includes the elements from the third and fourth rows and the fifth and sixth columns:
\[\begin{bmatrix}
5 & 28 \\
15 & 19 \\
\end{bmatrix} \]
<br/>
**Max Value: 28**
<br/>
<br/>
**Region (4,0) to (5,1)**
This region includes the elements from the fifth and sixth rows and the first and second columns:
\[\begin{bmatrix}
13 & 36 \\
47 & 29 \\
\end{bmatrix} \]
<br/>
**Max Value: 47**
<br/>
<br/>
**Region (4,2) to (5,3)**
This region includes the elements from the fifth and sixth rows and the third and fourth columns:
\[\begin{bmatrix}
24 & 48 \\
4 & 6 \\
\end{bmatrix} \]
<br/>
**Max Value: 48**
<br/>
<br/>
**Region (4,4) to (5,5)**
This region includes the elements from the fifth and sixth rows and the fifth and sixth columns:
\[\begin{bmatrix}
3 & 20 \\
42 & 31 \\
\end{bmatrix} \]
<br/>
**Max Value: 42**
<br/>
<br/>

### The final Max Pool output:
\
\begin{bmatrix}
38 & 33 & 34 \\
25 & 40 & 28 \\
47 & 48 & 42 \\
\end{bmatrix}

In [3]:
import numpy as np

# Given input matrix
I = np.array([
    [17, 14, 21, 0, 30, 32],
    [1, 38, 16, 33, 34, 23],
    [12, 7, 26, 18, 5, 28],
    [22, 25, 11, 40, 15, 19],
    [13, 36, 24, 48, 3, 20],
    [47, 29, 4, 6, 42, 31]
])

# Max pooling with 2x2 filter and stride of 2
pool_size = 2
stride = 2

output_dim = ((I.shape[0] - pool_size) // stride + 1, (I.shape[1] - pool_size) // stride + 1)
pooled_output = np.zeros(output_dim)

for i in range(0, I.shape[0], stride):
    for j in range(0, I.shape[1], stride):
        if i + pool_size <= I.shape[0] and j + pool_size <= I.shape[1]:
            pooled_output[i//stride, j//stride] = np.max(I[i:i+pool_size, j:j+pool_size])

print("Max Pooled Output:")
print(pooled_output)


Max Pooled Output:
[[38. 33. 34.]
 [25. 40. 28.]
 [47. 48. 42.]]


In [1]:
import torch
import torch.nn as nn

class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()
        self.conv1 = nn.Conv2d(in_channels=3, out_channels=16, kernel_size=5, stride=2, padding=2)  # SAME padding
        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
        self.conv2 = nn.Conv2d(in_channels=16, out_channels=32, kernel_size=3, stride=1, padding=1)  # SAME padding
        self.fc1 = nn.Linear(4 * 4 * 32, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = self.pool(x)
        x = self.conv2(x)
        x = self.pool(x)
        x = x.view(-1, 4 * 4 * 32)  # Flatten
        x = self.fc1(x)
        return x

# Instantiate the model
model = SimpleCNN()
print(model)


SimpleCNN(
  (conv1): Conv2d(3, 16, kernel_size=(5, 5), stride=(2, 2), padding=(2, 2))
  (pool): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (conv2): Conv2d(16, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (fc1): Linear(in_features=512, out_features=10, bias=True)
)


In [3]:
import tensorflow as tf
from tensorflow.keras import layers, models

model = models.Sequential()
model.add(layers.Conv2D(16, (5, 5), strides=2, padding='same', input_shape=(32, 32, 3)))
model.add(layers.MaxPooling2D((2, 2), strides=2))
model.add(layers.Conv2D(32, (3, 3), strides=1, padding='same'))
model.add(layers.MaxPooling2D((2, 2), strides=2))
model.add(layers.Flatten())
model.add(layers.Dense(10))

# Compile the model
# model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Display the model summary
model.summary()


Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 16, 16, 16)        1216      
                                                                 
 max_pooling2d_2 (MaxPoolin  (None, 8, 8, 16)          0         
 g2D)                                                            
                                                                 
 conv2d_3 (Conv2D)           (None, 8, 8, 32)          4640      
                                                                 
 max_pooling2d_3 (MaxPoolin  (None, 4, 4, 32)          0         
 g2D)                                                            
                                                                 
 flatten_1 (Flatten)         (None, 512)               0         
                                                                 
 dense_1 (Dense)             (None, 10)               

<br/>
<br/>
<br/>

# Question 5:

Consider a simple RNN with a linear activate function for one output defined by these equations, for



st = c * (wx * xt + wrec * st-1)
<br/>
and
y=c(wh * sT)




Assume

wx = 4 , wrec = 2, wh = 3, s0 = 0, c = 2, T = 3, and x = (1, 2, 2).

In [13]:
# Answering Question 5 through Python Code - Option # 1
def rnn_output(x, wx, wrec, wh, s0, c):
  st = s0
  for xt in x:
    st = c* (wx*xt + wrec*st)
  y = c* (wh*st) # This shouldn't be in the for loop as it is not looping with xt
  return y

# Providing values as mentioned in Question 5 to this function
x =[1,2,2]
wx = 4
wrec = 2
wh = 3
s0 = 0
c = 2
output = rnn_output(x,wx,wrec, wh,s0,c)
print(f'OUTPUT: {output}')


OUTPUT: 1248


<br/>

### Answering Question 5 through LaTex format:
Solving this problem to calculate st at each time step and then final state to compute the output y.

**Based on the below equations:**

st = c * (wx * xt + wrec * st-1)
<br/>
and
y = c * (wh * sT)

**With the assumption of the these values:**
wx = 4 , wrec = 2, wh = 3, s0 = 0, c = 2, T = 3, and x = (1, 2, 2).

**Lets Compute the state st at each time step:**
<br/>

**Time Step 1 (t=1):**

*   s1 = c * (wx * x1 + wrec * s0) = 2 * (4*1+2*0) = 8
<br/>

**Time Step 2 (t=2):**


*   s2 = c * (wx * x2 + wrec * s1) = 2 * (4*2+2*8) = 48
<br/>

**Time Step 3 (t=3):**


*   s2 = c * (wx * x3 + wrec * s2) = 2 * (4*2+2*48) = 208
<br/>

**Now Lets calculate the output y:**

y = c * (wh * sT)

y = 2 * (3 * 208) = 1248

**The Final output y is "1248"**













# Question 10:

Given the following input sequence with 3 vectors and 4 elements for each vector, and the self-attention mechanism parameters:

Input Sequence:
[
[
4
,
2
,
1
,
5
]
,
[
2
,
1
,
4
,
2
]
,
[
4
,
7
,
0
,
2
]
]
[[4,2,1,5],[2,1,4,2],[4,7,0,2]]

Parameters:
### Weight Matrices (for demonstration purposes)
\[ Query Weight = \begin{bmatrix}
0.5 & 0 & 0 & 0 \\
0 & 0.3 & 0 & 0 \\
0 & 0 & 0.2 & 0 \\
0 & 0 & 0 & 0.1
\end{bmatrix} \]
\[ Key Weight = \begin{bmatrix}
0.4 & 0 & 0 & 0 \\
0 & 0.6 & 0 & 0 \\
0 & 0 & 0.2 & 0 \\
0 & 0 & 0 & 0.8
\end{bmatrix} \]
\[ Value Weight = \begin{bmatrix}
0.1 & 0 & 0 & 0 \\
0 & 0.9 & 0 & 0 \\
0 & 0 & 0.6 & 0 \\
0 & 0 & 0 & 0.3
\end{bmatrix} \]

(Hint: Attention

(Q,K,V)= softmax
(QKT/dk))V , where Q=XWQ, K=XWK , V=XWV , X is the input, Wi is the weight matrix, dk is the scale factor.)

Calculate the Query vector, Key vector and Value vector given the above input and parameters.

(i) Show your calculation process for the first vector [4,2,1,5].

(ii) Show the result for the whole input sequence.

<br/>

### **Answer 10 (i)**
<br/>

### Show your calculation process for the first vector [4, 2, 1, 5]

**First Vector Calculation [4, 2, 1, 5]:**


Q_1 =
\begin{bmatrix}
4 & 2 & 1 & 5
\end{bmatrix}

\[ Query Weight = \begin{bmatrix}
0.5 & 0 & 0 & 0 \\
0 & 0.3 & 0 & 0 \\
0 & 0 & 0.2 & 0 \\
0 & 0 & 0 & 0.1
\end{bmatrix} \]
\[ Key Weight = \begin{bmatrix}
0.4 & 0 & 0 & 0 \\
0 & 0.6 & 0 & 0 \\
0 & 0 & 0.2 & 0 \\
0 & 0 & 0 & 0.8
\end{bmatrix} \]
\[ Value Weight = \begin{bmatrix}
0.1 & 0 & 0 & 0 \\
0 & 0.9 & 0 & 0 \\
0 & 0 & 0.6 & 0 \\
0 & 0 & 0 & 0.3
\end{bmatrix} \]


Calculate each element of
Q1 =

*   Q1[0] = 4 * 0.5 + 2 * 0 + 1 * 0 + 5 * 0 = 2
*   Q1[1] = 4 * 0 + 2 * 0.3 + 1 * 0 + 5 * 0 = 0.6

*   Q1[2] = 4 * 0 + 2 * 0 + 1 * 0.2 + 5 * 0 = 0.2
*   Q1[3] = 4 * 0 + 2 * 0 + 1 * 0 + 5 * 0.1 = 0.5
<br/>
<br/>

Calculate each element of
K1 =

*   K1[0] = 4 * 0.4 + 2 * 0 + 1 * 0 + 5 * 0 = 1.6
*   K1[1] = 4 * 0 + 2 * 0.6 + 1 * 0 + 5 * 0 = 1.2

*   K1[2] = 4 * 0 + 2 * 0 + 1 * 0.2 + 5 * 0 = 0.2
*   K1[3] = 4 * 0 + 2 * 0 + 1 * 0 + 5 * 0.8 = 4
<br/>
<br/>

Calculate each element of
V1 =

*   V1[0] = 4 * 0.1 + 2 * 0 + 1 * 0 + 5 * 0 = 0.4
*   V1[1] = 4 * 0 + 2 * 0.9 + 1 * 0 + 5 * 0 = 1.8

*   V1[2] = 4 * 0 + 2 * 0 + 1 * 0.6 + 5 * 0 = 0.6
*   V1[3] = 4 * 0 + 2 * 0 + 1 * 0 + 5 * 0.3 = 0.15
<br/>

**So, Q1 = [ 2, 0.6, 0.2, 5]**
<br/>

**So, K1 = [ 1.6, 1.2, 0.2, 4]**
<br/>

**So, V1 = [ 0.4, 1.8, 0.6, 0.15]**




***The above calculation is shown here to show the calculation process for the first vector and the below code is to show the result for the whole input sequence for Question 10***

In [34]:
import torch
import numpy as np
# Input sequence
X = np.array([
    [4, 2, 1, 5],
    [2, 1, 4, 2],
    [4, 7, 0, 2]
])

# Weight matrices
W_Q = np.array([
    [0.5, 0, 0, 0],
    [0, 0.3, 0, 0],
    [0, 0, 0.2, 0],
    [0, 0, 0, 0.1]
])

W_K = np.array([
    [0.4, 0, 0, 0],
    [0, 0.6, 0, 0],
    [0, 0, 0.2, 0],
    [0, 0, 0, 0.8]
])

W_V = np.array([
    [0.1, 0, 0, 0],
    [0, 0.9, 0, 0],
    [0, 0, 0.6, 0],
    [0, 0, 0, 0.3]
])

Q = np.dot(X, W_Q)
K = np.dot(X, W_K)
V = np.dot(X, W_V)

print(f"First Query Vector: {Q[0]}")
print(f"First Key Vector: {K[0]}")
print(f"First Value Vector: {V[0]}")

print("Query Vector:")
print(Q)

print("Key Vector:")
print(K)

print("Value Vector:")
print(V)

First Query Vector: [2.  0.6 0.2 0.5]
First Key Vector: [1.6 1.2 0.2 4. ]
First Value Vector: [0.4 1.8 0.6 1.5]
Query Vector:
[[2.  0.6 0.2 0.5]
 [1.  0.3 0.8 0.2]
 [2.  2.1 0.  0.2]]
Key Vector:
[[1.6 1.2 0.2 4. ]
 [0.8 0.6 0.8 1.6]
 [1.6 4.2 0.  1.6]]
Value Vector:
[[0.4 1.8 0.6 1.5]
 [0.2 0.9 2.4 0.6]
 [0.4 6.3 0.  0.6]]


# Question 11:
Based on your calculations above

(i) Calculate the raw attention scores (you don’t need to apply the scale factor or the softmax function.)for the second vector of the input. Show your calculation process and result.

(ii) Then you apply the softmax function to this raw attention scores to get the attention score (you don’t need to apply the scale factor). Your result can contain the exponential term.

<br/>
<br/>
<br/>


**Answering the question 11 (i):**
<br/>
<br/>

**Showing my calculation process and result through manual for Question 11(i) and below cell will display the answer in code:**

Detailed Dot Product Calculation:


Raw Attention Scores=Q2⋅K Transpose

For the second input vector:

Q2 = [1, 0.3, 0.8, 0.2]
<br/>

Key matrix:

\[ K =\begin{bmatrix}
1.6 & 1.2 & 0.2 & 4.0 \\
0.8 & 0.6 & 0.8 & 1.6 \\
1.6 & 4.2 & 0 & 1.6 \\
\end{bmatrix}

]  


Transposed Key matrix:

\[K Transpose =
\begin{bmatrix}
1.6 & 0.8 & 1.6 \\
1.2 & 0.6 & 4.2 \\
0.2 & 0.8 & 0 \\
4.0 & 1.6 & 1.6
\end{bmatrix}

]  

**Computing the dot products:**
<br/>


*   Raw Attention Scores [0] = (1 * 1.6) + (0.3 * 1.2) + (0.8 * .2) +(0.2 * 4.0) = 2.92
*   Raw Attention Scores [1] = (1 * 0.8) + (0.3 * 0.6) + (0.8 * 0.8) +(0.2 * 1.6) = 1.94

*   Raw Attention Scores [2] = (1 * 1.6) + (0.3 * 4.2) + (0.8 * 0) +(0.2 * 1.6) = 3.18

**Raw Attention Scores: [ 2.92, 1.94, 3.18]**










In [19]:
#Answering the question 11(i) through coding:

# To calcualate the raw attention score for the second vector without the applying the scale factor (dk)

# The score is calculated by taking the dot product of the Query Vector (Q2) with Key Vector (K) of the respective word we're scoring.
# So, if we're processing the self-attenction for the word in position #1.
# The first score would be the dot product of Q2 and K1 and second score would be the dot product of Q2 and K2 and so on.
# Mathematically it requires the mulitiplying the query vector with the transponse of Key Matrix

# Therefore Attention scores = Q2. K.Transpose



raw_attention_scores = np.dot(Q[1], K.T)
# Q[1] is the second query vector
# K.T is the transpose of the Key Matrix

print("Raw Attention Scores and this is answering the question 11(i):")
print(raw_attention_scores)



Raw Attention Scores and this is answering the question 11(i):
[2.92 1.94 3.18]


**Answering the question 11 (ii):**
<br/>
<br/>

**Showing my calculation process and result through manual for Question 11(ii) and below cell will display the answer in code:**

\begin{align*}
\text{Softmax}(2.92) &= \frac{e^{2.92}}{e^{2.92} + e^{1.94} + e^{3.18}} = 18.541 / 49.545 = 0.3742 \\
\text{Softmax}(1.94) &= \frac{e^{1.94}}{e^{2.92} + e^{1.94} + e^{3.18}} = 6.958 / 49.545 = 0.1404 \\
\text{Softmax}(3.18) &= \frac{e^{3.18}}{e^{2.92} + e^{1.94} + e^{3.18}} = 24.046 / 49.545 = 0.4853
\end{align*}

<br/>
<br/>

In [29]:
#Answering the question 11(ii) through coding:

import torch
import torch.nn.functional as F

# Apply softmax and using the softmax from pytorch.
# Not applying the

Softmax = F.softmax(torch.tensor(raw_attention_scores), dim=-1)

print("Attention Scores and this is answering the question 11(ii):")
print(Softmax)

Attention Scores and this is answering the question 11(ii):
tensor([0.3742, 0.1404, 0.4853], dtype=torch.float64)


# Question 12:

Calculate the final representation of the second vector after this self-attention mechanism, using the
𝑄
Q vector,
𝐾
K vector,
𝑉
V vector and the attention value you calculated above. For the attention score, you can just use the raw attention scores. Show your calculation process and the result.

<br/>
<br/>

### Answer 12:

**Showing my calculation process and result through manual for Question 12 and below cell will display the answer in code:**

\[
V = \begin{bmatrix}
0.4 & 1.8 & 0.6 & 1.5 \\
0.2 & 0.9 & 2.4 & 0.6 \\
0.4 & 6.3 & 0 & 0.6
\end{bmatrix}
\]

Softmax values = \\begin{bmatrix}
0.3742 & 0.1404 & 0.4853
\end{bmatrix}

<br/>


**Calculate each component:**


*   0.3742 * 0.4 = 0.1432
*   0.1404 * 0.2 = 0.0280

*   0.4853 * 0.4 = 0.1941

*   0.3742 * 1.8 = 0.6735
*   0.1404 * 0.9 = 0.1263

*   0.4853 * 6.3 = 3.0573
*   0.3742 * 0.6 = 0.2245
*   0.1404 * 2.4 = 0.3369

*   0.4853 * 0 = 0
*   0.3742 * 1.5 = 0.5613
*   0.1404 * 0.6 = 0.0842

*   0.4853 * 0.6 = 0.2911

<br/>


**Summing these values for each component gives the final representation:**

\begin{bmatrix}
0.1432 + 0.0280 + 0.1941 & 0.6735 + 0.1263 + 3.0573 & 0.2245 + 0.3369 + 0 & 0.5613 + 0.0842 + 0.2911
\end{bmatrix}

<br/>

**Final Representation of the Second Vector:**

\\begin{bmatrix}
0.3653 & 3.8571 & 0.5614 & 0.9366
\end{bmatrix}



In [30]:
#Answering the question 12 through coding:

output = np.dot(Softmax, V)

print("\nFinal Representation of the second vector:")
print(output)


Final Representation of the second vector:
[0.37191039 3.85760073 0.56160598 0.93679595]
